server/src/main/java/io/envoyproxy/controlplane/server/limits/ManualFlowControl.java

```diff
@@ -0,0 +1,78 @@
+package io.envoyproxy.controlplane.server.limits;
+
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.grpc.stub.ServerCallStreamObserver;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ManualFlowControl implements FlowControl<DiscoveryResponse> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ManualFlowControl.class);
+
+  private final long streamId;
+
+  private final ServerCallStreamObserver<DiscoveryResponse> serverCallStreamObserver;
+
+  private final AtomicBoolean observerWasReady = new AtomicBoolean(false);
+
+  private final RequestLimiter requestLimiter;
+
+  /**
+   * Flow Control implementation that considers back-pressure and request limiting.
+   * @param streamId ID of the stream
+   * @param serverCallStreamObserver the response observer which allows to control the client's readiness
+   * @param requestLimiter a synchronous limiter
+   */
+  public ManualFlowControl(long streamId,
+                           ServerCallStreamObserver<DiscoveryResponse> serverCallStreamObserver,
+                           RequestLimiter requestLimiter) {
+    this.streamId = streamId;
+    this.serverCallStreamObserver = serverCallStreamObserver;
+    this.requestLimiter = requestLimiter;
+  }
+
+  @Override
+  public void streamOpened() {
+    serverCallStreamObserver.disableAutoInboundFlowControl();
+    serverCallStreamObserver.setOnReadyHandler(() -> {
+      if (serverCallStreamObserver.isReady() && observerWasReady.compareAndSet(false, true)) {
+        LOGGER.debug("[{}] stream ready", streamId);
+        serverCallStreamObserver.request(1);
```

Comment:
Should you be acquiring a rate limit here? Conceptually it seems like each permit maps to a request?

Reply:
Well, this is the situation when the consumer's (client reading responses) HTTP2 buffer gains enough space to consume more messages. In this case it's not really a message **yet**. The permit is performed **after** processing the request, not before.